Validator check for cyclical imports

hsorby opened the pull request:
Closes #796.

hsorby committed:
Working through adding validator checks for cyclical imports.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

hsorby committed:
Correct paths to import resources.  Remove debugging statements.

hsorby committed:
Add test for units which are deeper than the first entry.

hsorby commented:
Supercedes #799.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

hsorby committed:
Apply clang format.

hsorby committed:
Merge branch 'validator-check-for-cyclical-imports' of github.com:hsorby/libcellml into validator-check-for-cyclical-imports

hsorby committed:
Rename some utility functions; getUnitsIndexInModel -> indexOf, getVariableIndexInComponent -> indexOf, reverseEngineerIndexStack -> indexStackOf.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

agarny committed:
Some minor cleaning up.

agarny committed:
Merge branch 'validator-check-for-cyclical-imports' of https://github.com/hsorby/libcellml into pr/837

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2843,3 +2843,172 @@ TEST(Validator, duplicateIdAll)
+        " - component 'i_am_cyclic' is imported from 'c2' in 'circularImport_2.cellml';\n"
+        " - component 'c2' is imported from 'c3' in 'circularImport_3.cellml';\n"
+        " - component 'c3' is imported from 'i_am_cyclic' in 'circularImport_1.cellml'; and\n"
+        " - component 'i_am_cyclic' is imported from 'c2' in 'circularImport_2.cellml'.";
```
It's sufficient to say that we have `i_am_cyclic` -> `c2` -> `c3` -> `i_am_cyclic`. No need to say that we have `i_am_cyclic` -> `c2` -> `c3` -> `i_am_cyclic` -> `c2`.

IOW, I would remove the second `component 'i_am_cyclic' is imported from 'c2' in 'circularImport_2.cellml'`.

hsorby replied:
3565c3f

agarny committed:
Some minor cleaning up.

agarny committed:
Removed unneeded CellML codes.

agarny reviewed (dismissed):
Would like to see the cyclic message to report about `a` -> `b` -> `c` -> `a`, and not `a` -> `b` -> `c` -> `a` -> `b`.

agarny committed:
Partly revert a3d4dbc6.

We need this for the importSecondGenComponentWithInvalidName validator test.

hsorby committed:
Add debug print for HistoryEntry.

hsorby committed:
Shorten history of loop report to the bare loop.

hsorby committed:
Merging changes from 'origin/validator-check-for-cyclical-imports' into validator-check-for-cyclical-imports fixing conflicts.

hsorby commented:
This PR is now ready for review.

hsorby committed:
Fixes for clang format.

nickerso commented on tests/importer/importer.cpp:
```diff
@@ -56,10 +56,9 @@ TEST(Importer, warningCircularImportReferencesComponent)
-        " - component 'i_am_cyclic' is imported from 'c2' in 'circularImport_2.cellml'.";
+        " - the component 'i_am_cyclic' is referencing a component 'c2' in the model 'circularImport2' imported from 'circularImport_2.cellml';\n"
+        " - the component 'c2' is referencing a component 'c3' in the model 'circularImport3' imported from 'circularImport_3.cellml'; and\n"
+        " - the component 'c3' is referencing a component 'i_am_cyclic' in the model 'circularImport1' imported from 'circularImport_1.cellml'.";
```
this set of messages is not defining a circular error as there is nothing to say that `c3` is importing the same `i_am_cyclic` component...

agarny replied:
?? We are saying that "Cyclic dependencies were found when attempting to resolve a component in the model 'circularImport1'" and that `c3` references "'i_am_cyclic' in the model 'circularImport1' imported from 'circularImport_1.cellml'". So...?

That aside, I would reword the error message, which I find too verbose:
```
Cyclic dependencies were found when attempting to resolve component 'i_am_cyclic' in 'circularImport_1.cellml'. The dependency loop is:
 - component 'i_am_cyclic' references component 'c2' in 'circularImport_2.cellml';
 - component 'c2' references component 'c3' in 'circularImport_3.cellml';
 - component 'c3' references component 'i_am_cyclic' in 'circularImport_1.cellml'.
```

(There is only one model in a CellML file, so no need to mention it. To mention the CellML filename is sufficient.)

hsorby replied:
We seem to have one vote each way here.

I see `I_am_cyclic -> c2 -> c3 -> i_am_cyclic` which shows a loop.  Do you disagree with that @nickerso?

nickerso replied:
LOL - I can't read. Didn't see the first line in the error message as it was not highlighted by Github as a difference - so thanks to @agarny in the above comment making that clearer.

I'd prefer to see the href's resolved to absolute URLs as that is the information I actually need to debug circular imports as the filenames on their own are pretty meaningless, but that can be a future fix if needed.

agarny replied:
> I'd prefer to see the href's resolved to absolute URLs as that is the information I actually need to debug circular imports as the filenames on their own are pretty meaningless, but that can be a future fix if needed.

Agreed with absolute URLs. One might, for some reasons, have one or several CellML files with the same name.

nickerso commented on tests/importer/importer.cpp:
```diff
@@ -76,10 +75,9 @@ TEST(Importer, warningCircularImportReferencesUnits)
-        " - units 'i_am_cyclic' is imported from 'u2' in 'circularUnits_2.cellml'.";
+        " - the units 'i_am_cyclic' is referencing units 'u2' in the model 'circularImport2' imported from 'circularUnits_2.cellml';\n"
+        " - the units 'u2' is referencing units 'u3' in the model 'circularImport3' imported from 'circularUnits_3.cellml'; and\n"
+        " - the units 'u3' is referencing units 'i_am_cyclic' in the model 'circularImport1' imported from 'circularUnits_1.cellml'.";
```
again, without that final message this is not describing the circular imports...

nickerso commented on tests/validator/validator.cpp:
```diff
@@ -2848,10 +2848,9 @@ TEST(Validator, circularImportReferencesComponent)
-        " - component 'i_am_cyclic' is imported from 'c2' in 'circularImport_2.cellml'.";
+        " - the component 'i_am_cyclic' is referencing a component 'c2' in the model 'circularImport2' imported from 'circularImport_2.cellml';\n"
+        " - the component 'c2' is referencing a component 'c3' in the model 'circularImport3' imported from 'circularImport_3.cellml'; and\n"
+        " - the component 'c3' is referencing a component 'i_am_cyclic' in the model 'circularImport1' imported from 'circularImport_1.cellml'.";
```
not circular...

nickerso commented on tests/validator/validator.cpp:
```diff
@@ -2872,10 +2871,9 @@ TEST(Validator, circularImportReferencesUnits)
-        " - units 'i_am_cyclic' is imported from 'u2' in 'circularUnits_2.cellml'.";
+        " - the units 'i_am_cyclic' is referencing units 'u2' in the model 'circularImport2' imported from 'circularUnits_2.cellml';\n"
+        " - the units 'u2' is referencing units 'u3' in the model 'circularImport3' imported from 'circularUnits_3.cellml'; and\n"
+        " - the units 'u3' is referencing units 'i_am_cyclic' in the model 'circularImport1' imported from 'circularUnits_1.cellml'.";
```
not circular

nickerso reviewed (changes requested):
Some of the new error reporting is not reporting sufficient information to define the circular imports. Not sure why some of them do...

hsorby commented:
Is the model name more important than the file the model is from?

nickerso commented:
> Is the model name more important than the file the model is from?

currently the model name is used for nothing. The URL of the import source is the important information to debug circular imports.

nickerso commented on src/api/libcellml/validator.h:
```diff
@@ -59,6 +59,19 @@ class LIBCELLML_EXPORT Validator: public Logger
+     *
+     * Validate the given @p model and its encapsulated entities using the
+     * CellML 2.0 Specification, after resolving the imports against the given @p baseLocation.
+     * Any errors will be logged in the @c Validator.
```
This seems to suggest that the full model hierarchy will be validated, but implementation currently validates the top-level model only and then adds in importer issues (e.g., circular imports) but doesn't seem to validate all the imported models? (maybe still a work in progress?)

hsorby replied:
I don't think we want to be validating all the imported models. We only want to validate the bits we are importing surely?

nickerso replied:
Exactly, but currently we are not validating the imported bits at all, other than issues that arise during the resolution process. At least at an initial look. If this is ready for review, then I can dive deeper into the behaviour to see if it is actually doing what I think it should be...

hsorby committed:
Simplify looping issue messages.

hsorby committed:
Add full resolved path to error messages for resolving import loops.  Tidy implementation of importing components and units.

hsorby committed:
Add missing generator_p.h from header files list.

hsorby committed:
Add import history tracking to isResolved methods for Units and Components.

hsorby committed:
Add some tests showing that xml ids are not validated and imported units and components are not validated.

hsorby committed:
Add implementation for detecting invalid xml attribute ids.

hsorby committed:
Implement validating models with imports resolved.

hsorby committed:
Get all the tests in validator passing.

hsorby committed:
Correct error in replace units where parent was not set.  Remove printIssues calls.

hsorby committed:
Remove validateModel method with base location.

hsorby committed:
Remove unused function indexOf(units).

hsorby committed:
Remove unused method checkUnitsForCycles in validator.cpp.

hsorby committed:
Move split function from validator to utilities.

hsorby committed:
Add documentation to new functions and methods.

hsorby committed:
Initialise mQ in Component.cpp.

hsorby committed:
Apply some code fixes generated from clang warnings.

hsorby committed:
Fix clang errors in characterBreakdown.

hsorby committed:
Add documentaiton to new function convertTextToUint32. Cast to uint32_t in convertTextToUint32.

hsorby committed:
Use auto when static casting.

hsorby committed:
Fix clang format errors.

hsorby committed:
Fix more clang format errors.

hsorby commented:
This is ready for a deeper review now.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

nickerso committed:
Merge branch 'main' into validator-check-for-cyclical-imports

nickerso commented on tests/resources/annotator/invalid_ids_on_ervery_element.cellml:
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
is this "typo in filename" file actually used anywhere?

hsorby replied:
I don't believe it is.

nickerso replied:
b0d2865

nickerso commented on tests/units/units.cpp:
```diff
@@ -2687,3 +2687,35 @@ TEST(Units, unknownUnitsScalingFactorIncompatible)
     EXPECT_EQ(0.0, scaling);
 }
+
+TEST(Units, circularImport)
```
not sure what this test has to do with `circularImport`...

nickerso reviewed (dismissed):
apart from perhaps mistakenly adding an unused test CellML model, this is looking pretty good to me.

nickerso committed:
Delete invalid_ids_on_ervery_element.cellml

Unused file accidentally included in PR.

hsorby committed:
Rename test to make its purpose clearer.

hsorby committed:
Merge branch 'validator-check-for-cyclical-imports' of github.com:hsorby/libcellml into validator-check-for-cyclical-imports

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

agarny commented on src/component.cpp:
```diff
@@ -51,6 +52,8 @@ struct Component::ComponentImpl
     bool equalVariables(const ComponentPtr &other) const;
     bool equalResets(const ComponentPtr &other) const;
+
+    bool isResovledWithHistory(ImportHistory &history) const;
```
`isResovledWithHistory` ---> `isResolvedWithHistory`

hsorby replied:
487c2a7

agarny commented on src/component.cpp:
```diff
@@ -41,6 +41,7 @@ namespace libcellml {
  */
 struct Component::ComponentImpl
 {
+    Component *mQ = nullptr;
```
is `mQ` a standard name to designate the instance of the API class? `mQ` certainly doesn't mean anything to me. If anything, I might use something like `mOwner`?

hsorby replied:
`mQ` is the return of `mPimpl` I wouldn't say it was a standard though.

hsorby replied:
Following QT here, We have used `mValidator` when in the pimpl for the Validator class for instance.

hsorby replied:
I don't think `mOwner` conveys the correct meaning to what it is.

agarny replied:
Not sure why `mOwner` is not fine since it's definitely the owner of an instance of `ComponentImpl`. Anyway, I would certainly be happy with being consistent with what has already been done for the `Validator` class, i.e. use `mComponent` instead of `mQ`.

hsorby replied:
40ce4fa

agarny reviewed (commented):
Just a couple of quick comments at this stage. (Will finish my review "later"...)

hsorby committed:
Track the source of errors through the import hierarchy.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

hsorby committed:
Use empty() instead of size() when testing for non-zero string length.

hsorby committed:
Remove extra blank lines in validator.cpp tests.

hsorby commented:
This has everything that I want in it now.  Ready for a final deep review.

agarny commented on src/internaltypes.h:
```diff
@@ -55,10 +55,19 @@ using ResetWeakPtr = std::weak_ptr<Reset>; /**< Type definition for weak reset p
 using ConnectionMap = std::map<VariablePtr, VariablePtr>; /**< Type definition for a connection map.*/
 
+using HistoryEntry = std::tuple<std::string, std::string, std::string>;
+using HistoryList = std::vector<HistoryEntry>;
```
For consistency (with `ImportHistory`/`ImportHistoryEntry`), we ought to rename `HistoryList` to `History`.

hsorby replied:
a32ae1f

agarny commented on src/units.cpp:
```diff
@@ -152,6 +158,79 @@ bool Units::UnitsImpl::isBaseUnit(const std::string &name) const
+    if (isStandardUnitName(unitsName)) {
+        standardUnitCheck = isBaseUnit(unitsName);
+    }
+    return mQ->unitCount() == 0 && standardUnitCheck;
```
To improve readability, I would replace with `return (mQ->unitCount() == 0) && standardUnitCheck;`.

hsorby replied:
574cdaf

agarny commented on src/utilities.h:
```diff
@@ -708,4 +711,36 @@ bool equalEntities(const EntityPtr &owner, const std::vector<EntityPtr> &entitie
+                                   const HistoryList &history,
+                                   const std::string &action);
+
+// Would be nice to add documentation to these.
```
Would delete that comment and, instead, create a GitHub issue for it.

hsorby replied:
#857.

agarny replied:
Happy with issue #857, but just in case: still need to delete that comment.

agarny commented on src/validator.cpp:
```diff
@@ -167,7 +169,7 @@ struct Validator::ValidatorImpl
      * to track repeated component names.
      */
-    void validateComponentTree(const ModelPtr &model, const ComponentPtr &component, std::vector<std::string> &componentNames);
+    void validateComponentTree(const ModelPtr &model, const ComponentPtr &component, std::vector<std::string> &componentNames, HistoryList &history, std::vector<ModelPtr> &visitedModels);
```
Need to update the documentation.

hsorby replied:
1ff5455

agarny commented on src/validator.cpp:
```diff
@@ -143,19 +157,7 @@ struct Validator::ValidatorImpl
-     * @param component The imported component to validate.
-     */
-    void validateImportedComponent(const ComponentPtr &component) const;
+    void validateComponent(const ComponentPtr &component, HistoryList &history, std::vector<ModelPtr> &visitedModels) const;
```
Need to update the documentation. (One of the reasons I wouldn't bother with documenting "internal" methods.)

hsorby replied:
1ff5455

agarny commented on src/validator.cpp:
```diff
@@ -229,9 +232,8 @@ struct Validator::ValidatorImpl
-     * @param unitsNames A vector list of the name attributes of the @p units and its siblings.
      */
-    void validateUnitsUnit(size_t index, const UnitsPtr &units, const std::vector<std::string> &unitsNames) const;
+    void validateUnitsUnit(size_t index, const UnitsPtr &units, HistoryList &history, const std::vector<ModelPtr> &modelsVisited) const;
```
Need to update the documentation.

hsorby replied:
1ff5455

agarny commented on src/validator.cpp:
```diff
@@ -397,8 +380,61 @@ struct Validator::ValidatorImpl
+     *
+     * @return @c true if the @p names have already been reported in a cyclic issue, @c false otherwise.
+     */
+    bool checkCycleForDuplications(std::vector<std::string> names) const;
```
Replace `std::vector<std::string>` with `Strings`. Also ought to be done eveywhere in the codebase, but maybe best done as part of another GitHub issue?

hsorby replied:
#856.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1628,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+}
+
+/**
+ * @brief Test to determine if name is a valid xml name.
```
XML, not xml.

hsorby replied:
1ff5455

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1628,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+        return true;
+    }
+    // "-" | "." | [0-9] | #xB7 | [#x0300-#x036F] | [#x203F-#x2040]
+    return ((0x30U <= nameChar && nameChar <= 0x39U) || (nameChar == 0x2DU) || (nameChar == 0x2EU) || (nameChar == 0xB7U) || (0x0300U <= nameChar && nameChar <= 0x036FU) || (0x203FU <= nameChar && nameChar <= 0x2040U));
```
Readability:
```c++
return (((0x30U <= nameChar) && (nameChar <= 0x39U)) 
       || (nameChar == 0x2DU)
       || (nameChar == 0x2EU)
       || (nameChar == 0xB7U)
       || ((0x0300U <= nameChar) && (nameChar <= 0x036FU))
       || ((0x203FU <= nameChar) && (nameChar <= 0x2040U)));
```

hsorby replied:
I'll have to check with clang format.  Clang format specified the current layout.

agarny replied:
Any update on this?

agarny replied:
Also, no need for the surrounding brackets.

agarny replied:
> I'll have to check with clang format. Clang format specified the current layout.

Just checked: we can have the following:
```c++
return ((0x30U <= nameChar) && (nameChar <= 0x39U))
       || (nameChar == 0x2DU)
       || (nameChar == 0x2EU)
       || (nameChar == 0xB7U)
       || ((0x0300U <= nameChar) && (nameChar <= 0x036FU))
       || ((0x203FU <= nameChar) && (nameChar <= 0x2040U));
```

agarny commented on tests/importer/importer.cpp:
```diff
@@ -56,10 +56,12 @@ TEST(Importer, warningCircularImportReferencesComponent)
+                                      " - component 'c2' references component 'c3' in '"
+        + resourcePath("importer/") + "circularImport_3.cellml'; and\n"
+                                      " - component 'c3' references component 'i_am_cyclic' in '"
+        + resourcePath("importer/") + "circularImport_1.cellml'.";
```
Can't this be reformatted as (same elsewhere):
```c++
        " - component 'i_am_cyclic' references component 'c2' in '" + resourcePath("importer/") + "circularImport_2.cellml';\n"
        " - component 'c2' references component 'c3' in '" + resourcePath("importer/") + "circularImport_3.cellml'; and\n"
        " - component 'c3' references component 'i_am_cyclic' in '" + resourcePath("importer/") + "circularImport_1.cellml'.";
```
Indentation for "component `c2`" and "component `c3`" makes it confusing.

hsorby replied:
I'll check with clang format I think that is where this format came from.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -99,6 +99,26 @@ TEST(Validator, invalidCellMLIdentifiersWithSpecificationHeading)
+TEST(Validator, invalidElementIdAttribute)
+{
+    const std::vector<std::string> expectedIssues = {
+        "Model 'valid_name' does not have a valid 'id' attribute, '993-++$@'.",
```
I would reformat the error message to read:
```
Model 'valid_name' does not have a valid 'id' attribute ('993-++$@').
```

agarny replied:
We agreed that the exact styling of error messages needs to be reviewed, so resolving this conversation.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2229,6 +2287,63 @@ TEST(Validator, unitComplexCycle)
+{
+    // Same test as unitComplexCycle except the order of the units
+    // added to the model is different.  This changes the order
+    // that the loop is detected  in.
```
Delete extra space between "detected" and "in".

hsorby replied:
1ff5455

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2229,6 +2287,63 @@ TEST(Validator, unitComplexCycle)
 
+TEST(Validator, unitComplexCycleChangeOrderOfUnits)
+{
+    // Same test as unitComplexCycle except the order of the units
+    // added to the model is different.  This changes the order
```
Would be good to refactor this example so that it uses the same code as in `unitComplexCycle()` and use a boolean (for instance) to determine the order in which the units should be added.

hsorby replied:
af1b101

hsorby committed:
Update W3 id error to use new UnitsItem.

hsorby committed:
Rename mQ to mComponent and mUnits in respective compilation units.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+                                      " - component 'c3' references component 'i_am_cyclic' in '"
+        + resourcePath("importer/") + "circularImport_1.cellml'; and\n"
+                                      " - component 'i_am_cyclic' references component 'c2' in '"
+        + resourcePath("importer/") + "circularImport_2.cellml'.";
```
Clearly not critical, but for consistency, it would be nice to move this before the declaration of `errorMessagesValidator`.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+        "Import of component 'c' has an invalid URI in the xlink:href attribute.",
+        "The attempt to resolve imports with the model at '"
+            + resourcePath("importer/")
+            + "i am broken and invalid.cellml' failed: the file could not be opened.",
```
Would put this on one line (i.e. one "sentence" per line). (Same in other places.)

agarny replied:
We should agree on the best way to report issues that provide additional information. For instance, above, we have "Model 'valid_name' does not have a valid 'id' attribute, '993-++$@'.", i.e. provide additional information after a comma while here we do it after a semicolon. Personally, I would just have:
 - "Model 'valid_name' does not have a valid 'id' attribute ('993-++$@')."; and
 - "The attempt to resolve imports with the model at '" + resourcePath("importer/") + "i am broken and invalid.cellml' failed (the file could not be opened).".

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+TEST(Validator, importComponentWithInvalidName)
+{
+    const std::vector<std::string> errorMessages = {
+        "Import of component 'c' has an invalid URI in the xlink:href attribute.",
```
Should read "... in the 'xlink:href' attribute.", i.e. use single quotes around "xlink:href".

In fact, there are many other cases where we should use single quotes around the attribute name (check in `src/validator.cpp`):
 - "name attribute" ---> "'name' attribute";
 - "xlink:href attribute" ---> "'xlink:href' attribute";
 - "component_ref attribute" ---> "'component_ref' attribute";
 - "units_ref attribute" ---" 'units_ref' attribute";
 - "units attribute" ---> "'units' attribute";
 - "interface attribute" ---> "'interface' attribute"; and
 - "cellml:units attribute" ---> "'cellml:units' attribute".

hsorby replied:
I think this should be a separate issue if it is widespread.

agarny replied:
See issue #854.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    const std::string errorMessageValidator =
+        "Imported component 'c' is not valid because:\n"
+        "  -> Component 'c' imported as 'c' from 'invalid_import_url.cellml' has an error:\n"
+        "  -> Import of component 'c' has an invalid URI in the xlink:href attribute.";
```
The use of "because", arrows, capital letters, "has an error:", etc. breaks our current convention and, in fact, makes this error message very confusing. At this stage, and assuming I understand the error message right, I would probably have something like:
```
Component 'c' is imported from component 'c' in 'invalid_import_url.cellml' which is invalid (it has an invalid URI in the 'xlink:href' attribute).
```

hsorby replied:
I agree (partly).  The reason for it is that you don't know the depth of the error.  It could be four or five layers of imports deep.  I think we need to discuss this reporting of imports errors in more depth.

agarny replied:
Agreed, something that needs to be discussed.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+}
+
+TEST(Validator, importSecondGenComponentWithInvalidName)
```
Is the name of the test correct? Where is the invalid name in the error message?

hsorby replied:
ea3e649

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    const std::string errorMessageValidator =
+        "Imported units 'i_am_bad' is not valid because:\n"
+        "  -> Units 'ps' imported as 'i_am_bad' from 'units_library.cellml' has an error:\n"
+        "  -> Units reference 'seconds' in units 'ps' is not a valid reference to a local units or a standard unit type.";
```
I would rephrase to:
```
Units 'ps' is imported from units 'i_am_bad' in 'units_library.cellml' which is invalid (units item 'seconds' is not a valid reference to a local units or a standard unit type).
```

(Note that we shouldn't use "units reference" but "units item".)

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    const std::string errorMessage =
+        "Imported units 'i_am_bad' is not valid because:\n"
+        "  -> Units 'ps2' imported as 'i_am_bad' from 'units_library.cellml' has an error:\n"
+        "  -> Prefix 'plinco' of a unit referencing 'second' in units 'ps2' is not a valid integer or an SI prefix.";
```
I would rephrase to:
```
Units 'ps2' is imported from units 'i_am_bad' in 'units_library.cellml' which is invalid (prefix 'plinco' of a unit item 'second' is not a valid integer or an SI prefix).
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+TEST(Validator, importInvalidUnitsReference)
+{
+    const std::string errorMessage =
+        "Imported units 'i_am_bad' refers to units 'ps3' which does not appear in 'units_library'.";
```
We shouldn't mention the name of the imported model, but its file name (same in test `importInvalidComponentReference`):
```
Units 'i_am_bad' is imported from units 'ps3' which does not appear in 'units_library.cellml'.
```

hsorby replied:
64cd2ed

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    const std::string errorMessage2 =
+        "Imported component 'imported_component' is not valid because:\n"
+        "  -> Component 'invalid_component' imported as 'imported_component' from 'component_that_is_invalid.cellml' has an error:\n"
+        "  -> Component 'invalid_component' does not have a valid 'id' attribute, '8ioisfje+'.";
```
I would rephrase to:
```
Component 'invalid_component' is imported from component 'imported_component' in 'component_that_is_invalid.cellml' which is invalid (it does not have a valid 'id' attribute ('8ioisfje+')).
```

(Double parentheses are not great, but I could personally live with them.)

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+        "Imported component 'component1' is not valid because:\n"
+        "  -> Component 'component2' imported as 'component1' from 'model2.cellml' has an error:\n"
+        "  -> Component 'component3' imported as 'component2' from 'model3.cellml' has an error:\n"
+        "  -> Variable 'time' in component 'component3' has a units reference 'seconds' which is neither standard nor defined in the parent model.";
```
Very confusing indeed...

hsorby committed:
Rename HistoryList to History.

hsorby committed:
Documentation corrections.

hsorby committed:
Fix readability issue in src/units.cpp.

hsorby committed:
Add documentation to some of the functions in utilities.h.

hsorby committed:
Refactor unitsCycleComplex.

hsorby committed:
Change name of test importSecondGenComponentWithInvalidName to importSecondGenComponentWithInvalidURI.

agarny commented on src/validator.cpp:
```diff
@@ -229,9 +236,10 @@ struct Validator::ValidatorImpl
+     * @param modelsVisited The list of visited models.
      */
-    void validateUnitsUnit(size_t index, const UnitsPtr &units, const std::vector<std::string> &unitsNames) const;
+    void validateUnitsUnit(size_t index, const UnitsPtr &units, History &history, const std::vector<ModelPtr> &modelsVisited) const;
```
I imagine this method should be renamed to `validateUnitsUnitsItem()`?

hsorby replied:
2d559a7

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,465 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+}
+
+TEST(Validator, importSecondGenComponentWithInvalidURI)
```
To respect our convention: `importSecondGenComponentWithInvalidUri()`.

hsorby replied:
b904e22

hsorby committed:
Rename test importSecondGenComponentWithInvalidURI to importSecondGenComponentWithInvalidUri.

hsorby committed:
Rename validateUnitsUnit to validateUnitsUnitsItem.

hsorby committed:
Report filename not model name when reporting import errors.

hsorby committed:
Report filename not model name for component import errors.

nickerso committed:
Demonstrate a potential bug in the testing for cyclical imports.

Adding a new test which has two models which import the same common units definitions and then one
of the models also imports the other model. This results in an cyclical import error reported when
the importer resolves the imports. The error message is wrong in that it refers to components as
units; and also wrong in that as far as I can tell there is no cycle in the imported units.

The test models added are the minimal demonstration of this that I've reduced from an actual model.
Reducing further (e.g., fewer variables, or only importing a subset of the units) results in different
behaviour, which is possibly a different issue.

hsorby committed:
Merge pull request #13 from nickerso/cyclical-imports

Demonstrate a potential bug in the testing for cyclical imports.

hsorby committed:
Get all tests passing for Importer and Validator.

hsorby committed:
Get the implementation for cyclical imports to work with the desired outputs.

hsorby committed:
Remove debug statements from utilities.cpp.

hsorby committed:
Remove old makeIssueCyclicDependency function.

hsorby committed:
Rename makeIssueCyclicDependency2 to makeIssueCyclicDependency.

hsorby committed:
Define a string to use for the model origin reference value.

hsorby committed:
Improve coverage.  Add test showing a C shape import hierarchy.

hsorby committed:
Rename variable hh to history.

hsorby committed:
Convert isResolvedWithHistory to use ImportTrack.

hsorby committed:
Convert units isResolvedWithHistory to use ImportTrack.

hsorby committed:
Convert isBaseUnitWithHistory to use ImportTrack.

hsorby committed:
Rename ImportTrack to History.

hsorby committed:
Remove ImportHistoryEntry and ImportHistory definitions.

hsorby committed:
Give validator test a better name.

hsorby committed:
Rename ImportStep to HistoryEpoch.

hsorby committed:
Move implementation of HistoryEpoch into implementation file.

hsorby committed:
Add a createHistoryEpoch utility function.

hsorby committed:
Tidy up use of createHistoryEpoch.

hsorby committed:
Rename createImportStep to createHistoryEpoch.

hsorby committed:
Add documentation for internal functions; importeeModelUrl, checkForImportCycles.

hsorby committed:
Add documentation for HistoryEpoch class.

hsorby committed:
Add (missing) & in importeeModelUrl function.

hsorby committed:
Correct clang tidy errors.

hsorby committed:
Use std::any_of instead of loop.

hsorby committed:
Apply clang format fixes.

hsorby committed:
Apply clang format fixes.

agarny commented on Testing/Temporary/CTestCostData.txt:
```diff
@@ -0,0 +1 @@
+---
```
What is this file about? Same for `Testing/Temporary/LastTest.log`.

hsorby replied:
File added by accident, removed in 032d033.

agarny replied:
Yes, it was an "old" comment.

agarny commented on src/debug.cpp:
```diff
@@ -28,14 +28,19 @@ limitations under the License.
 {
-    for (auto tt : history) {
-        Debug() << std::get<0>(tt) << " " << std::get<1>(tt) << " " << std::get<2>(tt);
+    for (auto h : history) {
```
`const auto &h` rather than `auto h`?

agarny commented on src/debug.cpp:
```diff
@@ -28,14 +28,19 @@ limitations under the License.
-    for (auto tt : history) {
-        Debug() << std::get<0>(tt) << " " << std::get<1>(tt) << " " << std::get<2>(tt);
+    for (auto h : history) {
+        printHistoryEpoch(h);
```
Do we really need `printHistoryEpoch()`? It's only used here and I would probably not expect us to use it for debugging (we would rather use `printHistory()`, I imagine?).

hsorby replied:
No, when you are checking for cycles you often want to print out the entry you are about to check.

agarny replied:
Ok, fair enough.

agarny commented on src/internaltypes.h:
```diff
@@ -21,10 +21,16 @@ limitations under the License.
+#include "libcellml/model.h"
+#include "libcellml/units.h"
+
+#include <iostream>
```
Should be moved between lines 19 and 20.

hsorby replied:
Shouldn't be there I will remove it.

agarny commented on src/importer.cpp:
```diff
@@ -99,7 +97,28 @@ std::vector<ImportSourcePtr>::const_iterator Importer::ImporterImpl::findImportS
+        }
+    }
+
+    return std::string(ORIGIN_MODEL_REF);
```
Replace with:
```c++
return ORIGIN_MODEL_REF;
```

hsorby committed:
Add a badly designed model to show the behaviour of detecting import cycles with the importer.

hsorby committed:
Fix clang format erros.

hsorby committed:
Add printImportLibrary debugging function.

hsorby committed:
Fix incorrect name of component_ref in 'layer1/importing_bad_design_z_import_hierarchy.cellml' model.

hsorby committed:
Add counterpoint test for the bad z import hierarchy.

hsorby committed:
Fix bug found when adding z offset import hierarchy test.

agarny commented on src/debug.cpp:
```diff
@@ -492,4 +497,17 @@ void printAst(const AnalyserEquationAstPtr &ast)
+void printImportLibrary(const ImportLibrary &importlibrary)
+{
+    Debug() << " == Import library ==";
+    for ( const auto &entry : importlibrary) {
```
Might want our ClangFormat target to format this file too? I mean, there should be no space before `const`.

agarny commented on src/units.cpp:
```diff
@@ -139,11 +139,17 @@ struct Units::UnitsImpl
+
+    bool isBaseUnitWithHistory(History &history, const UnitsConstPtr &units) const;
+
+    bool isResolvedWithHistory(History &history, const UnitsConstPtr &units) const;
```
Replace with `isResolvedUnitWithHistory()`?

hsorby replied:
I don't think so the underlying question is `isResolved` and the extension is with history.

agarny replied:
Fair enough.

nickerso commented on Testing/Temporary/CTestCostData.txt:
```diff
@@ -0,0 +1 @@
+---
```
this file shouldn't be here, right?

hsorby replied:
032d033

nickerso commented on Testing/Temporary/LastTest.log:
```diff
@@ -0,0 +1,3 @@
+Start testing: May 12 08:13 NZST
```
another file that shouldn't be added, right?

hsorby replied:
032d033

nickerso commented on src/importer.cpp:
```diff
@@ -123,122 +142,107 @@ bool Importer::ImporterImpl::checkUnitsForCycles(const ModelPtr &origModel, cons
+    auto model = units->importSource()->model();
+    if (model == nullptr) {
+        auto issue = Issue::create();
+        issue->setDescription("Units '" + units->name() + "' requires a model imported from '" + units->importSource()->url() + "' which is not available in the importer.");
```
should this now be the `resolvingUrl(units->importSource())` ?

hsorby replied:
560943b

nickerso commented on src/importer.cpp:
```diff
@@ -123,122 +142,107 @@ bool Importer::ImporterImpl::checkUnitsForCycles(const ModelPtr &origModel, cons
+    auto importedUnits = model->units(units->importReference());
+    if (importedUnits == nullptr) {
+        auto issue = Issue::create();
+        issue->setDescription("Units '" + units->name() + "' imports units named '" + units->importReference() + "' from the model imported from '" + units->importSource()->url() + "'. The units could not be found.");
```
`resolvingUrl(units->importSource())` ?

hsorby replied:
560943b

nickerso commented on src/internaltypes.h:
```diff
@@ -21,10 +21,16 @@ limitations under the License.
 
 namespace libcellml {
 
+const std::string ORIGIN_MODEL_REF = "this";
```
should this be an invalid URI to ensure that, for valid models, there can't be any confusion when testing for url's that equal `ORIGIN_MODEL_REF` ? maybe `:this` ?

hsorby replied:
985f460

nickerso commented on tests/importer/importer.cpp:
```diff
@@ -50,16 +50,19 @@ TEST(Importer, noWarningDiamondImport)
-        " - component 'c2' is imported from 'c3' in 'circularImport_3.cellml';\n"
-        " - component 'c3' is imported from 'i_am_cyclic' in 'circularImport_1.cellml'; and\n"
-        " - component 'i_am_cyclic' is imported from 'c2' in 'circularImport_2.cellml'.";
+        " - component 'i_am_cyclic' specifies an import from 'this' to '"
```
shouldn't these messages include the name of the component being imported from the given URL? Its not a circular import if you are importing independent components from an already imported model....is it?

hsorby replied:
I take 
```
Any CellML infoset imported, directly or indirectly, by the imported CellML infoset MUST NOT be semantically equivalent to the importing CellML infoset
```
to mean that I can't make any loops irrespective of what I am extracting.

agarny replied:
Only comment left. @nickerso, are you ok with @hsorby's reply?

nickerso reviewed (commented):
A few questions to answer before a final look through this one...

hsorby committed:
Remove files incorrectly added.

hsorby committed:
Remove a whole bunch of cmake_minimum_required.

hsorby committed:
Use full resolving url in importer error messages.

hsorby committed:
Set the name of ORIGIN_MODEL_REF to :this:.

hsorby committed:
Fix clang format errors.

agarny commented on src/units.cpp:
```diff
@@ -152,6 +158,79 @@ bool Units::UnitsImpl::isBaseUnit(const std::string &name) const
+                history.push_back(h);
+                if (model->hasUnits(mUnits->importReference())) {
+                    auto importedUnits = model->units(mUnits->importReference());
+                    return importedUnits->mPimpl->isBaseUnitWithHistory(history, importedUnits); // Call isBaseUnit recursively until unit is no longer an import
```
Could put the comment above this line? Otherwise, missing full stop.

agarny commented on src/units.cpp:
```diff
@@ -152,6 +158,79 @@ bool Units::UnitsImpl::isBaseUnit(const std::string &name) const
+        }
+    }
+    return resolved;
+}
```
I would rework the code to read:
```c++
bool Units::UnitsImpl::isResolvedWithHistory(History &history, const UnitsConstPtr &units) const
{
    if (!mUnits->isImport()) {
        return true;
    }

    auto model = mUnits->importSource()->model();
    if (model == nullptr) {
        return false;
    }

    auto importedUnits = model->units(mUnits->importReference());
    if (importedUnits == nullptr) {
        return false;
    }

    auto h = createHistoryEpoch(units, importeeModelUrl(history, mUnits->importSource()->url()));
    if (checkForImportCycles(history, h)) {
        return false;
    }

    if (importedUnits->isImport()) {
        history.push_back(h);

        return importedUnits->mPimpl->isResolvedWithHistory(history, importedUnits);
    }

    for (size_t u = 0; (u < importedUnits->unitCount()); ++u) {
        std::string reference;
        std::string prefix;
        std::string id;
        double exponent;
        double multiplier;
        importedUnits->unitAttributes(u, reference, prefix, exponent, multiplier, id);
        if (isStandardUnitName(reference)) {
            continue;
        }

        auto childUnits = model->units(reference);
        if (childUnits == nullptr) {
            return false;
        }

        history.push_back(h);

        if (!childUnits->mPimpl->isResolvedWithHistory(history, childUnits)) {
            return false;
        }
    }

    return true;
}
```

agarny commented on src/internaltypes.h:
```diff
@@ -55,10 +61,78 @@ using ResetWeakPtr = std::weak_ptr<Reset>; /**< Type definition for weak reset p
 using ConnectionMap = std::map<VariablePtr, VariablePtr>; /**< Type definition for a connection map.*/
 
+class HistoryEpoch; /**< Forward declaration of import step class. */
+using HistoryEpochPtr = std::shared_ptr<HistoryEpoch>; /**< Type definition for shared import step pointer. */
```
"import step class|pointer"?

agarny replied:
Also, might delete line 73 and move line 74 before line 134?

agarny commented on src/internaltypes.h:
```diff
@@ -55,10 +61,78 @@ using ResetWeakPtr = std::weak_ptr<Reset>; /**< Type definition for weak reset p
+    /**
+     * @brief Set the reference name for the imported entity.
+     *
+     * If the entity is imported set the reference name for it.
```
Need a comma before "set".

agarny commented on src/internaltypes.h:
```diff
@@ -55,10 +61,78 @@ using ResetWeakPtr = std::weak_ptr<Reset>; /**< Type definition for weak reset p
+    /**
+     * @brief Set the destination model to the resolving model from the imported entity.
+     *
+     * If the entity is imported set the destination model to the resolving model for it.
```
Need a comma before "set".

agarny commented on src/utilities.h:
```diff
@@ -521,6 +522,20 @@ Variable::InterfaceType determineInterfaceType(const VariablePtr &variable);
+ * @brief Split a string.
+ *
+ * Split the given string with the given delimiter.  If a delimiter is not given
+ * then the default delimiter is used.  The default delimiter is ';'.  If the
```
"... then ';' is used." rather than "... then the default delimiter is used.  The default delimiter is ';'."?

agarny commented on src/utilities.h:
```diff
@@ -521,6 +522,20 @@ Variable::InterfaceType determineInterfaceType(const VariablePtr &variable);
+ *
+ * Split the given string with the given delimiter.  If a delimiter is not given
+ * then the default delimiter is used.  The default delimiter is ';'.  If the
+ * delimiter is not found in @p content then a copy is returned.
```
"... a copy of @p content is returned."

agarny commented on src/utilities.h:
```diff
@@ -711,4 +728,110 @@ bool equalEntities(const EntityPtr &owner, const std::vector<EntityPtr> &entitie
+ *
+ * @return The history epoch.
+ */
+HistoryEpochPtr createHistoryEpoch(const UnitsConstPtr &units, const std::string &sourceUrl, const std::string &destinationUrl);
```
Why two methods? `destinationUrl` should be defaulted to `""`.

agarny commented on src/utilities.h:
```diff
@@ -711,4 +728,110 @@ bool equalEntities(const EntityPtr &owner, const std::vector<EntityPtr> &entitie
+ *
+ * @return The history epoch.
+ */
+HistoryEpochPtr createHistoryEpoch(const ComponentConstPtr &component, const std::string &sourceUrl, const std::string &destinationUrl);
```
Why two methods? `destinationUrl` should be defaulted to `""`.

agarny commented on src/utilities.cpp:
```diff
@@ -1365,4 +1381,86 @@ bool areEqual(const std::string &str1, const std::string &str2)
+HistoryEpochPtr createHistoryEpoch(const UnitsConstPtr &units, const std::string &sourceUrl, const std::string &destinationUrl)
+{
+    return std::make_shared<HistoryEpoch>(units, sourceUrl, destinationUrl);
+}
```
Replace with:
```c++
HistoryEpochPtr createHistoryEpoch(const UnitsConstPtr &units, const std::string &sourceUrl, const std::string &destinationUrl)
{
    auto h = std::make_shared<HistoryEpoch>(units, sourceUrl, destinationUrl);
    if (destinationUrl == "") {
        recordUrl(h, units);
    }
    return h;
}
```

agarny commented on src/utilities.cpp:
```diff
@@ -1365,4 +1381,86 @@ bool areEqual(const std::string &str1, const std::string &str2)
+HistoryEpochPtr createHistoryEpoch(const ComponentConstPtr &component, const std::string &sourceUrl, const std::string &destinationUrl)
+{
+    return std::make_shared<HistoryEpoch>(component, sourceUrl, destinationUrl);
+}
```
Replace with:
```c++
HistoryEpochPtr createHistoryEpoch(const ComponentConstPtr &component, const std::string &sourceUrl, const std::string &destinationUrl)
{
    auto h = std::make_shared<HistoryEpoch>(component, sourceUrl, destinationUrl);
    if (destinationUrl == "") {
        recordUrl(h, component);
    }
    return h;
}
```

agarny commented on src/validator.cpp:
```diff
@@ -93,6 +100,13 @@ IssuePtr makeIssueIllegalIdentifier(const std::string &name);
+ * @param name
+ * @return
+ */
+bool isValidW3IdName(const std::string &name);
```
- No description. In fact, would move the content of this method (and other methods it uses) here.
- Rename to `isValidXmlName()`.

agarny commented on src/validator.cpp:
```diff
@@ -397,8 +386,63 @@ struct Validator::ValidatorImpl
+     *
+     * @return @c true if the @p names have already been reported in a cyclic issue, @c false otherwise.
+     */
+    bool hasCycleAlreadyBeenReported(std::vector<std::string> names) const;
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -37,6 +39,11 @@ limitations under the License.
+/**
+ * Type definition for a list of issue descriptions.
+ */
+using IssuesList = std::vector<std::vector<std::string>>;
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -166,8 +170,10 @@ struct Validator::ValidatorImpl
+     * @param modelsVisited The list of visited models.
      */
-    void validateComponentTree(const ModelPtr &model, const ComponentPtr &component, std::vector<std::string> &componentNames);
+    void validateComponentTree(const ModelPtr &model, const ComponentPtr &component, std::vector<std::string> &componentNames, History &history, std::vector<ModelPtr> &modelsVisited);
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -322,7 +330,7 @@ struct Validator::ValidatorImpl
      * @param variableNames A @c vector list of the names of variables found within the @p component.
      */
-    void validateAndCleanMathCiCnNodes(XmlNodePtr &node, const ComponentPtr &component, const std::vector<std::string> &variableNames);
+    void validateAndCleanMathCiCnNodes(XmlNodePtr &node, const ComponentPtr &component, const std::vector<std::string> &variableNames) const;
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -549,54 +530,41 @@ void Validator::ValidatorImpl::validateUniqueName(const ModelPtr &model, const s
 }
 
-void Validator::ValidatorImpl::validateComponentTree(const ModelPtr &model, const ComponentPtr &component, std::vector<std::string> &componentNames)
+void Validator::ValidatorImpl::validateComponentTree(const ModelPtr &model, const ComponentPtr &component, std::vector<std::string> &componentNames, History &history, std::vector<ModelPtr> &modelsVisited)
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+
+    } else {
+        // Check for variables in this component.
+        std::vector<std::string> variableNames;
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+    handleErrorsFromImports(initialIssueCount, isOriginatingModel, "Component", componentName, history, component, nullptr);
+}
+
+std::set<std::string> namesInCycle(std::vector<std::string> allNames)
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+    return namesInCycle;
+}
+
+bool Validator::ValidatorImpl::hasCycleAlreadyBeenReported(std::vector<std::string> names) const
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+    if (checkForLocalCycles(history, h)) {
+        history.push_back(h);
+        std::string des;
+        std::vector<std::string> names;
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -1077,7 +1399,7 @@ void Validator::ValidatorImpl::validateAndCleanCiNode(const XmlNodePtr &node, co
 }
 
-void Validator::ValidatorImpl::validateAndCleanMathCiCnNodes(XmlNodePtr &node, const ComponentPtr &component, const std::vector<std::string> &variableNames)
+void Validator::ValidatorImpl::validateAndCleanMathCiCnNodes(XmlNodePtr &node, const ComponentPtr &component, const std::vector<std::string> &variableNames) const
```
Replace `std::vector<std::string>` with `Strings`.

agarny commented on src/validator.cpp:
```diff
@@ -176,9 +182,10 @@ struct Validator::ValidatorImpl
+     * @param modelsVisited The list of visited models.
      */
-    void validateUnits(const UnitsPtr &units, const std::vector<std::string> &unitsNames) const;
+    void validateUnits(const UnitsPtr &units, History &history, std::vector<ModelPtr> modelsVisited) const;
```
`std::vector<ModelPtr> &modelsVisited` rather than `std::vector<ModelPtr> modelsVisited`?

agarny commented on src/validator.cpp:
```diff
@@ -397,8 +386,63 @@ struct Validator::ValidatorImpl
+     * @param component The originating component if type is @c Component.
+     * @param units The originating units if type is @c Units.
+     */
+    void handleErrorsFromImports(size_t initialErrorCount, bool isOriginatingModel, const std::string &type, const std::string &name, const History &history, const ComponentPtr &component, const UnitsPtr &units) const;
```
We really ought to agree on a maximum line length. 219 characters...!

agarny commented on src/validator.cpp:
```diff
@@ -397,8 +386,63 @@ struct Validator::ValidatorImpl
+    /**
+     * @brief Deal with errors raised from imports.
+     *
+     * When an error is raised by an imported entity track back to the originating model and
```
Comma after "imported entity".

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+    const std::string skipThis = "Cyclic dependencies";
+    const std::string notOriginMarker = "NOT ORIGIN: ";
+    const std::string dataBoundaryMarker = "&";
+    const std::string dataSeparator = ";";
```
`static const...`

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
             validateVariable(variable, variableNames);
+            variableNames.push_back(variable->name());
+        }
+        // Check for resets in this component
```
Missing full stop.

agarny commented on src/validator.cpp:
```diff
@@ -694,7 +960,10 @@ void Validator::ValidatorImpl::validateUnitsUnit(size_t index, const UnitsPtr &u
     if (isCellmlIdentifier(reference)) {
-        if ((std::find(unitsNames.begin(), unitsNames.end(), reference) == unitsNames.end()) && (!isStandardUnitName(reference))) {
+        ModelPtr model = owningModel(units);
+        if (model->hasUnits(reference) && (!isStandardUnitName(reference))) {
```
Replace with:
```c++
if (model->hasUnits(reference) && !isStandardUnitName(reference)) {
```

agarny commented on src/validator.cpp:
```diff
@@ -694,7 +960,10 @@ void Validator::ValidatorImpl::validateUnitsUnit(size_t index, const UnitsPtr &u
+        ModelPtr model = owningModel(units);
+        if (model->hasUnits(reference) && (!isStandardUnitName(reference))) {
+            validateUnits(model->units(reference), history, modelsVisited);
+        } else if (!model->hasUnits(reference) && (!isStandardUnitName(reference))) {
```
Replace with:
```c++
} else if (!model->hasUnits(reference) && !isStandardUnitName(reference)) {
```

agarny commented on src/validator.cpp:
```diff
@@ -229,9 +236,10 @@ struct Validator::ValidatorImpl
+     * @param modelsVisited The list of visited models.
      */
-    void validateUnitsUnit(size_t index, const UnitsPtr &units, const std::vector<std::string> &unitsNames) const;
+    void validateUnitsUnitsItem(size_t index, const UnitsPtr &units, History &history, const std::vector<ModelPtr> &modelsVisited) const;
```
Remove `const` for `modelsVisited`.

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
 }
 
-void Validator::ValidatorImpl::validateUnitsUnit(size_t index, const UnitsPtr &units, const std::vector<std::string> &unitsNames) const
+void Validator::ValidatorImpl::validateUnitsUnitsItem(size_t index, const UnitsPtr &units, History &history, const std::vector<ModelPtr> &modelsVisited) const
```
Remove `const` for `modelsVisited`.

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+{
+    std::set<std::string> testNamesInCycle = namesInCycle(std::move(names));
+    bool found = false;
+    for (size_t i = 0; i < mValidator->issueCount() && !found; ++i) {
```
Replace with:
```c++
for (size_t i = 0; (i < mValidator->issueCount()) && !found; ++i) {
```

agarny commented on src/validator.cpp:
```diff
@@ -606,84 +574,382 @@ void Validator::ValidatorImpl::validateImportedComponent(const ComponentPtr &com
+
+    size_t unitsWithNameCount = 0;
+    size_t unitsWithImportSource = 0;
+    for (size_t i = 0; i < model->unitsCount() && modelsVisited.size() == 1; ++i) {
```
Replace with:
```c++
for (size_t i = 0; (i < model->unitsCount()) && (modelsVisited.size() == 1); ++i) {
```

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
 }
 
+/**
+ * @brief Test to determine if the @p name is a valid CellML identifier.
```
"the @p name" --> "@p name".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+/**
+ * @brief Test to determine if the @p name is a valid CellML identifier.
+ *
+ * Test to determine if the @p name is a valid CellML identifier.
```
"the @p name" --> "@p name".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
 }
 
+/**
+ * @brief Test to determine if character is a valid xml name start char.
```
"character" --> "@p startChar".
"xml name start char" --> "XML name start character".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+/**
+ * @brief Test to determine if character is a valid xml name start char.
+ *
+ * Start name char is defined here: https://www.w3.org/TR/xml11/#NT-NameStartChar.
```
"A start name character is..."

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+ *
+ * @param startChar The character to test.
+ *
+ * @return True if the character is in the allowed unicode ranges for a start character in a name.
```
"Unicode" rather than "unicode".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+bool isNameStartChar(uint32_t startChar)
+{
+    // ":" | [A-Z] | "_" | [a-z] | [#xC0-#xD6] | [#xD8-#xF6] | [#xF8-#x2FF] | [#x370-#x37D] | [#x37F-#x1FFF] | [#x200C-#x200D] | [#x2070-#x218F] | [#x2C00-#x2FEF] | [#x3001-#xD7FF] | [#xF900-#xFDCF] | [#xFDF0-#xFFFD] | [#x10000-#xEFFFF]
+    return ((startChar == 0x3AU) || (0x41U <= startChar && startChar <= 0x5AU) || (startChar == 0x5FU) || (0x61U <= startChar && startChar <= 0x7AU) || (0xC0U <= startChar && startChar <= 0xD6U) || (0xD8U <= startChar && startChar <= 0xF6U) || (0xF8U <= startChar && startChar <= 0x2FFU) || (0x370U <= startChar && startChar <= 0x37DU) || (0x37FU <= startChar && startChar <= 0x1FFFU) || (0x200CU <= ...
```
Replace with:
```c++
return (startChar == 0x3AU) || (0x41U <= startChar && startChar <= 0x5AU) || (startChar == 0x5FU) || (0x61U <= startChar && startChar <= 0x7AU) || (0xC0U <= startChar && startChar <= 0xD6U) || (0xD8U <= startChar && startChar <= 0xF6U) || (0xF8U <= startChar && startChar <= 0x2FFU) || (0x370U <= startChar && startChar <= 0x37DU) || (0x37FU <= startChar && startChar <= 0x1FFFU) || (0x200CU <= startChar && startChar <= 0x200DU) || (0x2070U <= startChar && startChar <= 0x218FU) || (0x2C00U <= startChar && startChar <= 0x2FEFU) || (0x3001U <= startChar && startChar <= 0xD7FFU) || (0xF900U <= startChar && startChar <= 0xFDCFU) || (0xFDF0U <= startChar && startChar <= 0xFFFDU) || (0x10000U <= startChar && startChar <= 0xEFFFFU);
```
i.e. no need for surrounding brackets.

agarny replied:
Should be replaced with:
```c++
return (startChar == 0x3AU) 
        || ((0x41U <= startChar) && (startChar <= 0x5AU)) 
        || (startChar == 0x5FU) 
        || ((0x61U <= startChar) && (startChar <= 0x7AU)) 
        || ((0xC0U <= startChar) && (startChar <= 0xD6U)) 
        || ((0xD8U <= startChar) && (startChar <= 0xF6U)) 
        || ((0xF8U <= startChar) && (startChar <= 0x2FFU)) 
        || ((0x370U <= startChar) && (startChar <= 0x37DU)) 
        || ((0x37FU <= startChar) && (startChar <= 0x1FFFU)) 
        || ((0x200CU <= startChar) && (startChar <= 0x200DU)) 
        || ((0x2070U <= startChar) && (startChar <= 0x218FU)) 
        || ((0x2C00U <= startChar) && (startChar <= 0x2FEFU)) 
        || ((0x3001U <= startChar) && (startChar <= 0xD7FFU)) 
        || ((0xF900U <= startChar) && (startChar <= 0xFDCFU)) 
        || ((0xFDF0U <= startChar) && (startChar <= 0xFFFDU)) 
        || ((0x10000U <= startChar) && (startChar <= 0xEFFFFU));
```

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+}
+
+/**
+ * @brief Convert a variable width codepoint to uint32 form.
```
"Convert @p text to a uint32 form."

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+ * @param text The @c std::string to convert.
+ * @param initialValue The starting point for the conversion.
+ *
+ * @return uint32 form of the @p text.
```
"... form of @p text."

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+/**
+ * @brief Breakdown a string into codepoints.
+ *
+ * Breakdown the @p text into a vector of codepoints as described by
```
"the @p text" --> "@p text".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+ *
+ * @param text The std::string to breakdown.
+ *
+ * @return A vector of @c uint32_t of the unicode character values.
```
"unicode" --> "Unicode".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+}
+
+/**
+ * @brief Breakdown a string into codepoints.
```
Here and elsewhere: "codepoints" --> "code points".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+    std::vector<uint32_t> breakdown;
+    std::vector<uint8_t> bitShifts = {24, 16, 8, 0};
+    for (size_t i = 0; i < text.length();) {
+        size_t codepointLength = 1;
```
"codepointLength" --> "codePointLength".

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+}
+
+/**
+ * @brief Test to determine if character is a valid XML name char.
```
"... if @p nameChar... XML name character."

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+/**
+ * @brief Test to determine if character is a valid XML name char.
+ *
+ * Name char is defined here: https://www.w3.org/TR/xml11/#NT-NameChar.
```
"An XML name character is..."

hsorby replied:
92fa27e

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+ *
+ * @param nameChar The character to test.
+ *
+ * @return True if the character is in the allowed unicode ranges for a name.
```
"... for an XML name character."

hsorby replied:
92fa27e

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+}
+
+/**
+ * @brief Test to determine if name is a valid XML name.
```
"... if @p name..."

hsorby replied:
92fa27e

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+/**
+ * @brief Test to determine if name is a valid XML name.
+ *
+ * Name is defined here: https://www.w3.org/TR/xml11/#NT-Name.
```
"An XML name is..."

hsorby replied:
92fa27e

agarny commented on src/validator.cpp:
```diff
@@ -1304,12 +1626,139 @@ Issue::ReferenceRule validateCellmlIdentifier(const std::string &name)
+ *
+ * @param name The @c std::string to test.
+ *
+ * @return True if the name is a valid xml name.
```
"... XML name."

hsorby replied:
92fa27e

agarny reviewed (commented):
Still have to go through the tests, but here goes for now...

agarny reviewed (commented):
Some more comments.

agarny commented on tests/resources/importer/common_units_import_1.xml:
```diff
@@ -0,0 +1,18 @@
+    <variable name="stimCurrent" initial_value="-10.0" units="uA_per_cmsq" interface="public"/>
+  </component>
+
+</model>
```
Not critical, but missing trailing line (same in some other test files).

agarny commented on tests/importer/importer.cpp:
```diff
@@ -50,16 +50,19 @@ TEST(Importer, noWarningDiamondImport)
+                                      " - component 'c2' specifies an import from '"
+        + resourcePath("importer/") + "circularImport_2.cellml' to '" + resourcePath("importer/") + "circularImport_3.cellml'; and\n"
+                                                                                                    " - component 'c3' specifies an import from '"
+        + resourcePath("importer/") + "circularImport_3.cellml' to '" + resourcePath("importer/") + "circularImport_1.cellml'.";
```
Replace with:
```c++
std::string("Cyclic dependencies were found when attempting to resolve a component in the model 'circularImport1'. The dependency loop is:\n")
+ " - component 'i_am_cyclic' specifies an import from ':this:' to '" + resourcePath("importer/") + "circularImport_2.cellml';\n"
+ " - component 'c2' specifies an import from '" + resourcePath("importer/") + "circularImport_2.cellml' to '" + resourcePath("importer/") + "circularImport_3.cellml'; and\n"
+ " - component 'c3' specifies an import from '" + resourcePath("importer/") + "circularImport_3.cellml' to '" + resourcePath("importer/") + "circularImport_1.cellml'.";
```
to make it much easier (for me at least!) to read.

hsorby replied:
Clang format says no.  I tried that already.

hsorby replied:
does the `std::string` make a difference for clang format here? rather than not having it.

agarny replied:
> Clang format says no. I tried that already.

I have tested the above code with ClangFormat 12 and it's all fine.

> does the `std::string` make a difference for clang format here? rather than not having it.

Don't know. All I know is that `std::string` is needed otherwise the compiler complains about us trying to add two `char*`.

agarny commented on tests/importer/importer.cpp:
```diff
@@ -76,10 +79,13 @@ TEST(Importer, warningCircularImportReferencesUnits)
+        + resourcePath("importer/") + "circularUnits_2.cellml' to '" + resourcePath("importer/") + "circularUnits_3.cellml'; and\n"
+                                                                                                   " - units 'u3' specifies an import from '"
+        + resourcePath("importer/") + "circularUnits_3.cellml' to '" + resourcePath("importer/") + "circularUnits_1.cellml'.";
+
```
Replace with:
```c++
std::string("Cyclic dependencies were found when attempting to resolve units in the model 'circularImport1'. The dependency loop is:\n")
+ " - units 'i_am_cyclic' specifies an import from ':this:' to '" + resourcePath("importer/") + "circularUnits_2.cellml';\n"
+ " - units 'u2' specifies an import from '" + resourcePath("importer/") + "circularUnits_2.cellml' to '" + resourcePath("importer/") + "circularUnits_3.cellml'; and\n"
+ " - units 'u3' specifies an import from '" + resourcePath("importer/") + "circularUnits_3.cellml' to '" + resourcePath("importer/") + "circularUnits_1.cellml'.";
```
to make it much easier (for me at least!) to read.

hsorby replied:
see above.

agarny replied:
Likewise. :)

agarny commented on tests/importer/importer.cpp:
```diff
@@ -1035,3 +1068,14 @@ TEST(Importer, removeAllModels)
+    EXPECT_EQ(size_t(0), parser->issueCount());
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(0), importer->issueCount());
+    EXPECT_EQ(size_t(0), importer->errorCount());
```
No need to test for `importer->errorCount()`.

agarny reviewed (commented):
Some momre comments.

agarny commented on tests/importer/model_flattening.cpp:
```diff
@@ -843,16 +843,21 @@ TEST(ModelFlattening, resolveFlattenCircularImportsComponents)
+                                             " - component 'shared' specifies an import from '"
+        + resourcePath("modelflattening/") + "circularImport1.cellml' to '" + resourcePath("modelflattening/") + "circularImport2.cellml'; and\n"
+                                                                                                                 " - component 'circular2' specifies an import from '"
+        + resourcePath("modelflattening/") + "circularImport2.cellml' to '" + resourcePath("modelflattening/") + "circularImport1.cellml'.";
```
Replace with:
```c++
const std::string resolveError =
    std::string("Cyclic dependencies were found when attempting to resolve a component in the model 'importExample2b'. The dependency loop is:\n")
    + " - component 'sideB' specifies an import from ':this:' to '" + resourcePath("modelflattening/") + "circularImport1.cellml';\n"
    + " - component 'shared' specifies an import from '" + resourcePath("modelflattening/") + "circularImport1.cellml' to '" + resourcePath("modelflattening/") + "circularImport2.cellml'; and\n"
    + " - component 'circular2' specifies an import from '" + resourcePath("modelflattening/") + "circularImport2.cellml' to '" + resourcePath("modelflattening/") + "circularImport1.cellml'.";

const std::string flattenError =
    std::string("Cyclic dependencies were found when attempting to flatten a component in the model 'importExample2b'. The dependency loop is:\n")
    + " - component 'sideB' specifies an import from ':this:' to '" + resourcePath("modelflattening/") + "circularImport1.cellml';\n"
    + " - component 'shared' specifies an import from '" + resourcePath("modelflattening/") + "circularImport1.cellml' to '" + resourcePath("modelflattening/") + "circularImport2.cellml'; and\n"
    + " - component 'circular2' specifies an import from '" + resourcePath("modelflattening/") + "circularImport2.cellml' to '" + resourcePath("modelflattening/") + "circularImport1.cellml'.";
```

hsorby replied:
has this been checked with clang format.

agarny replied:
Yes, using ClangFormat 12.

agarny commented on tests/importer/model_flattening.cpp:
```diff
@@ -877,16 +882,21 @@ TEST(ModelFlattening, resolveFlattenCircularImportsUnits)
+                                             " - units 'shared' specifies an import from '"
+        + resourcePath("modelflattening/") + "circularImport1units.cellml' to '" + resourcePath("modelflattening/") + "circularImport2units.cellml'; and\n"
+                                                                                                                      " - units 'circular2' specifies an import from '"
+        + resourcePath("modelflattening/") + "circularImport2units.cellml' to '" + resourcePath("modelflattening/") + "circularImport1units.cellml'.";
```
Replace with:
```c++
const std::string resolveError =
    std::string("Cyclic dependencies were found when attempting to resolve units in the model 'importExampleUnits'. The dependency loop is:\n")
    + " - units 'sideB' specifies an import from ':this:' to '" + resourcePath("modelflattening/") + "circularImport1units.cellml';\n"
    + " - units 'shared' specifies an import from '" + resourcePath("modelflattening/") + "circularImport1units.cellml' to '" + resourcePath("modelflattening/") + "circularImport2units.cellml'; and\n"
    + " - units 'circular2' specifies an import from '" + resourcePath("modelflattening/") + "circularImport2units.cellml' to '" + resourcePath("modelflattening/") + "circularImport1units.cellml'.";

const std::string flattenError =
    std::string("Cyclic dependencies were found when attempting to flatten units in the model 'importExampleUnits'. The dependency loop is:\n")
    + " - units 'sideB' specifies an import from ':this:' to '" + resourcePath("modelflattening/") + "circularImport1units.cellml';\n"
    + " - units 'shared' specifies an import from '" + resourcePath("modelflattening/") + "circularImport1units.cellml' to '" + resourcePath("modelflattening/") + "circularImport2units.cellml'; and\n"
    + " - units 'circular2' specifies an import from '" + resourcePath("modelflattening/") + "circularImport2units.cellml' to '" + resourcePath("modelflattening/") + "circularImport1units.cellml'.";
```

agarny commented on tests/importer/model_flattening.cpp:
```diff
@@ -929,7 +939,9 @@ TEST(ModelFlattening, resolveFlattenMissingModel)
-    const std::string e = "Component 'left' imports a component named 'pointyBit' from the model imported from 'diamond_point.cellml'. The component could not be found.";
+    const std::string e =
+        "Component 'left' imports a component named 'pointyBit' from the model imported from '"
+        + resourcePath("importer/") + "diamond_point.cellml'. The component could not be found.";
```
Replace with:
```c++
const std::string e =
    "Component 'left' imports a component named 'pointyBit' from the model imported from '" + resourcePath("importer/") + "diamond_point.cellml'. The component could not be found.";
```

agarny commented on tests/importer/model_flattening.cpp:
```diff
@@ -950,7 +962,9 @@ TEST(ModelFlattening, resolveFlattenMissingComponent)
-    const std::string e = "Units 'units1_imported' imports units named 'units1' from the model imported from 'units_source.cellml'. The units could not be found.";
+    const std::string e =
+        "Units 'units1_imported' imports units named 'units1' from the model imported from '"
+        + resourcePath("importer/") + "units_source.cellml'. The units could not be found.";
```
Replace with:
```c++
const std::string e =
    "Units 'units1_imported' imports units named 'units1' from the model imported from '" + resourcePath("importer/") + "units_source.cellml'. The units could not be found.";
```

agarny reviewed (commented):
Some more comments.

hsorby committed:
Format changes for name char.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2218,12 +2283,38 @@ TEST(Validator, unitComplexCycle)
+    // added to the model is different.  This changes the order
+    // that the loop is detected in.
+    const std::vector<std::string> expectedIssues = {
+        "Cyclic units exist: 'brotherFromAnotherMother' -> 'father' -> 'grandfather' -> 'brotherFromAnotherMother'",
```
I know we are going to review the format of our issues, but just in case:
 - Not great to start with "Cyclic units exist: ...". I would expect a proper sentence.
 - Missing full stop.

hsorby replied:
I think the first point should be added to our issue description document.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+                                      " - component 'c2' specifies an import from '"
+        + resourcePath("importer/") + "circularImport_2.cellml' to '" + resourcePath("importer/") + "circularImport_3.cellml'; and\n"
+                                                                                                    " - component 'c3' specifies an import from '"
+        + resourcePath("importer/") + "circularImport_3.cellml' to '" + resourcePath("importer/") + "circularImport_1.cellml'.";
```
Replace with:
```c++
const std::string errorMessageImporter =
    std::string("Cyclic dependencies were found when attempting to resolve a component in the model 'circularImport1'. The dependency loop is:\n")
    + " - component 'i_am_cyclic' specifies an import from ':this:' to '" + resourcePath("importer/") + "circularImport_2.cellml';\n"
    + " - component 'c2' specifies an import from '" + resourcePath("importer/") + "circularImport_2.cellml' to '" + resourcePath("importer/") + "circularImport_3.cellml'; and\n"
    + " - component 'c3' specifies an import from '" + resourcePath("importer/") + "circularImport_3.cellml' to '" + resourcePath("importer/") + "circularImport_1.cellml'.";
```

agarny replied:
What is `:this:`?

hsorby replied:
It is. the marker for the originating file that we don't have a filename for.

agarny replied:
We need (will have?) to think about a better way to convey that information. I could guess what `:this:` means, but I shouldn't have to guess. The error message should be crystal clear (and unanbiguous).

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+                                      " - units 'u2' specifies an import from '"
+        + resourcePath("importer/") + "circularUnits_2.cellml' to '" + resourcePath("importer/") + "circularUnits_3.cellml'; and\n"
+                                                                                                   " - units 'u3' specifies an import from '"
+        + resourcePath("importer/") + "circularUnits_3.cellml' to '" + resourcePath("importer/") + "circularUnits_1.cellml'.";
```
Replace with:
```c++
const std::string errorMessageImporter =
    std::string("Cyclic dependencies were found when attempting to resolve units in the model 'circularImport1'. The dependency loop is:\n")
    + " - units 'i_am_cyclic' specifies an import from ':this:' to '" + resourcePath("importer/") + "circularUnits_2.cellml';\n"
    + " - units 'u2' specifies an import from '" + resourcePath("importer/") + "circularUnits_2.cellml' to '" + resourcePath("importer/") + "circularUnits_3.cellml'; and\n"
    + " - units 'u3' specifies an import from '" + resourcePath("importer/") + "circularUnits_3.cellml' to '" + resourcePath("importer/") + "circularUnits_1.cellml'.";
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+                                                                                                   " - units 'u3' specifies an import from '"
+        + resourcePath("importer/") + "circularUnits_3.cellml' to '" + resourcePath("importer/") + "circularUnits_1.cellml'; and\n"
+                                                                                                   " - units 'i_am_cyclic' specifies an import from '"
+        + resourcePath("importer/") + "circularUnits_1.cellml' to '" + resourcePath("importer/") + "circularUnits_2.cellml'.";
```
Replace with:
```c++
const std::string errorMessageImporter =
    std::string("Cyclic dependencies were found when attempting to resolve units in the model 'circularImport1'. The dependency loop is:\n")
    + " - units 'i_am_duplicated' specifies an import from ':this:' to '" + resourcePath("importer/") + "circularUnits_2.cellml';\n"
    + " - units 'u2' specifies an import from '" + resourcePath("importer/") + "circularUnits_2.cellml' to '" + resourcePath("importer/") + "circularUnits_3.cellml';\n"
    + " - units 'u3' specifies an import from '" + resourcePath("importer/") + "circularUnits_3.cellml' to '" + resourcePath("importer/") + "circularUnits_1.cellml'; and\n"
    + " - units 'i_am_cyclic' specifies an import from '" + resourcePath("importer/") + "circularUnits_1.cellml' to '" + resourcePath("importer/") + "circularUnits_2.cellml'.";
```

agarny replied:
Would make it slightly easier to review code if we were to be consistent and always list `errorMessageImporter` first and then `errorMessagesValidator`, as we do in previous tests.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+                                                                                                    " - component 'c3' specifies an import from '"
+        + resourcePath("importer/") + "circularImport_3.cellml' to '" + resourcePath("importer/") + "circularImport_1.cellml'; and\n"
+                                                                                                    " - component 'i_am_cyclic' specifies an import from '"
+        + resourcePath("importer/") + "circularImport_1.cellml' to '" + resourcePath("importer/") + "circularImport_2.cellml'.";
```
Replace with:
```c++
const std::string errorMessageImporter =
    std::string("Cyclic dependencies were found when attempting to resolve a component in the model 'circularImport1'. The dependency loop is:\n")
    + " - component 'i_am_duplicated' specifies an import from ':this:' to '" + resourcePath("importer/") + "circularImport_2.cellml';\n"
    + " - component 'c2' specifies an import from '" + resourcePath("importer/") + "circularImport_2.cellml' to '" + resourcePath("importer/") + "circularImport_3.cellml';\n"
    + " - component 'c3' specifies an import from '" + resourcePath("importer/") + "circularImport_3.cellml' to '" + resourcePath("importer/") + "circularImport_1.cellml'; and\n"
    + " - component 'i_am_cyclic' specifies an import from '" + resourcePath("importer/") + "circularImport_1.cellml' to '" + resourcePath("importer/") + "circularImport_2.cellml'.";
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+        "The attempt to resolve imports with the model at '"
+            + resourcePath("importer/")
+            + "i am broken and invalid.cellml' failed: the file could not be opened.",
+    };
```
Replace with:
```c++
const std::vector<std::string> errorMessages = {
    "Import of component 'c' has an invalid URI in the xlink:href attribute.",
    "The attempt to resolve imports with the model at '" + resourcePath("importer/") + "i am broken and invalid.cellml' failed: the file could not be opened.",
};
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+    const std::string errorMessageImporter =
+        "The attempt to resolve imports with the model at '"
+        + resourcePath("importer/")
+        + "i am broken and invalid.cellml' failed: the file could not be opened.";
```
Replace with:
```c++
const std::string errorMessageImporter =
    "The attempt to resolve imports with the model at '" + resourcePath("importer/") + "i am broken and invalid.cellml' failed: the file could not be opened.";
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+TEST(Validator, importSecondGenComponentWithInvalidUri)
+{
+    const std::string errorMessageValidator =
+        "Imported component 'c' is not valid because:\n"
+        "  -> Component 'c' importing 'c' from 'invalid_import_url.cellml' has an error:\n"
+        "   - Import of component 'c' has an invalid URI in the xlink:href attribute.";
```
Really not a fan of that type of error message. I guess this falls under the to-be-reviewed category.

hsorby replied:
Certainly does.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+{
+    const std::string errorMessageImporter =
+        "Import of units 'i_am_bad' from '"
+        + resourcePath("importer/") + "units_library.cellml' requires units named 'ps', which relies on child units named 'seconds', which cannot be found.";
```
Replace with:
```c++
const std::string errorMessageImporter =
    "Import of units 'i_am_bad' from '" + resourcePath("importer/") + "units_library.cellml' requires units named 'ps', which relies on child units named 'seconds', which cannot be found.";
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+
+TEST(Validator, importInvalidComponentNotDirectlyDeterminedFromImport)
+{
+    const std::string errorMessage1 = "Component 'invalid_component' does not have a valid 'id' attribute, '8ioisfje+'.";
```
For consistency, may want to replace with:
```c++
const std::string errorMessage1 =
    "Component 'invalid_component' does not have a valid 'id' attribute, '8ioisfje+'.";
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+        " - component 'c1' specifies an import from ':this:' to '"
+        + resourcePath("importer/") + "component_library.cellml'; and\n"
+                                      " - component 'c1_imported' specifies an import from '"
+        + resourcePath("importer/") + "component_library.cellml' to '" + resourcePath("importer/") + "layer1/importing_bad_design_z_import_hierarchy.cellml'.";
```
Replace with:
```c++
const std::string errorMessage =
    std::string("Cyclic dependencies were found when attempting to resolve a component in the model 'import_component_from_library_and_another_component'. The dependency loop is:\n")
    + " - component 'c1' specifies an import from ':this:' to '" + resourcePath("importer/") + "component_library.cellml'; and\n"
    + " - component 'c1_imported' specifies an import from '" + resourcePath("importer/") + "component_library.cellml' to '" + resourcePath("importer/") + "layer1/importing_bad_design_z_import_hierarchy.cellml'.";
```

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(errorMessageValidator, validator->issue(0)->description());
+}
+
+TEST(Validator, highIndexUnitsImport)
```
The name of the test doesn't fit the test. What about `importedUnitBasedOnNonExistingUnit` instead?

hsorby replied:
1f87e46

agarny commented on tests/validator/validator.cpp:
```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(errorMessage, validator->error(0)->description());
+}
+
+TEST(Validator, cImportThatIllustratesBadPractice)
```
"cImport..."?

agarny replied:
It's not clear what this test actually does. Clearly, it's related to bad practice(s?), but which one(s?) are we talking about? Same story for the next two tests (i.e. `zImportThatIllustratesBadPractice` and `zOffsetImportThatIllustratesBadPractice`).

hsorby replied:
it is trying to say in a concise way how one might consider the type of import that is happening.  It is quite an edge case test so you really to have to consider the code in the test to understand exactly what is going on.

agarny reviewed (commented):
Some final comments... for now. :)

hsorby committed:
Improve documentationfor XML related functions.

hsorby committed:
Remove unwanted space in javascript/annotator.cpp

hsorby committed:
Rename highIndexUnitsImport to importedUnitBasedOnNonExistingUnit.

hsorby committed:
Change codepoint to code point.

hsorby committed:
Change unicode to Unicode.

hsorby committed:
Documentation fixes in src/validator.cpp.

hsorby committed:
Remove const from modelsVisited.

hsorby committed:
Add missing full stops.

agarny committed:
Merge branch 'main' into validator-check-for-cyclical-imports

agarny commented:
FTR: I have updated ClangFormat to version 12.0.0 on our CI machine (and now updating this PR so that our ClangFormat tests pass).

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make ClangFormat happy.

agarny committed:
Tests: fixed an expected issue.

agarny commented:
@hsorby, I have just created https://github.com/hsorby/libcellml/pull/14 for your PR. Otherwise, I am not sure what you have done in your recent commits, but the `Validator.unitSimpleCycle` test now expects three issues:
 - `Cyclic units exist: 'grandfather' -> 'child' -> 'father' -> 'grandfather'.`;
 - `Cyclic units exist: 'father' -> 'grandfather' -> 'child' -> 'father'.` and
 - `Cyclic units exist: 'child' -> 'father' -> 'grandfather' -> 'child'.`.

Something similar goes with tests `Validator.unitComplexCycle` and `Validator.unitComplexCycleChangeOrderOfUnits`.

hsorby committed:
Merge branch 'main' into validator-check-for-cyclical-imports

hsorby committed:
Fix problem with detecting loops in cyclic error descriptions.

hsorby committed:
Merge branch 'validator-check-for-cyclical-imports' of github.com:hsorby/libcellml into validator-check-for-cyclical-imports

hsorby committed:
Remove importer->errorCount().

hsorby committed:
Remove unnecessary #include <iostream>.

hsorby committed:
Use const .. & in history loop.

hsorby committed:
Remove std::string from ORIGIN_MODEL_REF.

hsorby committed:
Remove space in printImportLibrary for loop.

hsorby committed:
Refactor isResolvedwithHistory for Units.

hsorby committed:
Fix doxygen documentation in internaltypes.h.

hsorby committed:
Reorder declarations in internaltypes.h.

hsorby committed:
Add commas in HistoryEpoch method comments.

hsorby committed:
Improve documentation for split function.

hsorby committed:
Refactor createHistoryEpoch for component and units.

hsorby committed:
Move and rename isValidW3IdName to isValidXmlName.

hsorby committed:
Add comma after 'When an error is raised by an imported entity'.

hsorby committed:
Add static to const strings in handleErrorsFromImports.

hsorby committed:
Use typedef for std::vector<std::string>.

hsorby committed:
Use empty to test for empty string.

hsorby committed:
Ignore readability-function-cognitive-complexity in clang tidy.

hsorby committed:
Merge branch 'validator-check-for-cyclical-imports' into validator-check-for-cyclical-imports

hsorby committed:
Merge pull request #14 from agarny/validator-check-for-cyclical-imports

Validator check for cyclical imports

hsorby commented:
I think I have addressed all review comments except prettiness of error messages in tests which I will defer to another issue.  Can you please re-review. Ta.

hsorby commented:
Added an issue here for making test error strings more *readable* #862.

agarny committed:
Validator: slight reformatting to prevent long line.

I would normally use 80 characters per line max, but I believe some people might prefer 120 characters max per line. Something to discuss, but what is certain is that 219 characters is rather long.

agarny committed:
Validator: pass a parameter by reference rather than by copy.

agarny committed:
Tests: some reformatting of code to make it easier to read expected error messages.

agarny committed:
Tests: list importer error messages first and then validator error messages.

So that it's consistent throughout our tests.

agarny committed:
Tests: slight reformatting for consistency.

agarny reviewed (dismissed):
There were a few more pending comments, which I have addressed. The only one that remains is with regards `:this:`, but maybe that's something that could be addressed as part of our review of error messages.

agarny committed:
API documentation: some minor cleaning up.

agarny committed:
Validator: slight improvement to code readability.

agarny committed:
Removed trailing spaces here and there, and added trailing line wherever needed.

agarny committed:
Revert "Removed trailing spaces here and there, and added trailing line wherever needed."

This reverts commit 7c2b41df92e1ff43e0ae7d0f777987684c1b1969. This should be done as part of a separate GitHub issue.
